python/pyspark/sql/column.py

```diff
@@ -406,7 +406,13 @@ def substr(self, startPos, length):
         [Row(col=u'Ali'), Row(col=u'Bob')]
         """
         if type(startPos) != type(length):
-            raise TypeError("Can not mix the type")
+            raise TypeError(
+                "startPos and length must be the same type. "
+                "Got {startPos_t} and {length_t}, respectively."
```

Comment:
> -> startPos: {startPos_t}; length: {length_t}.

I think this suggestion basically includes the same information with the current change? I think you should explain why.

> BTW, why we do the type checking here, instead of doing it in the actual Scala impl of substr?

Are you saying exposing Java types in the error message is better or suggesting method signature in Scala impl of `substr` with the check logic?

> In addition, we do not support the mixed cases? For example, startPos is int, length is long.

In Python, it makes sense calling `int` in general I think. `long` and `int` are unified in Python 3 and this PR looks targeting only the exception message fix.